streaming/format/base/phaser.py

```diff
@@ -0,0 +1,103 @@
+            to not accept any uncheckable shard files you find cached locally upon init. If
+            ``Auto()``, falls back to ``True``. Defaults to ``Auto()``.
+        access (bool | Auto): Whether to cache or delete the very last phase of the file that its
+            format uses, without regard to whether it can be checked or not. This phase is the
+            final product of shard preparation. Set to ``True`` if you intend to use the shard in
+            any way, and to ``False`` for dry runs of shard preparation, clearing a dataset to just
+            the ``storage`` phase of shards, or other such purpose. If ``Auto``, falls back to
+            ``True``. Defaults to ``Auto()``.
+        zip (None | bool | Auto): Whether to cache or delete the ``zip`` phase of the file, if its
+            format uses it. If ``bool``, it overrides what the intended use cases would do. If
+            ``None``, it takes its value from from the intended use cases would do. If ``Auto``,
+            falls back to ``None``.
+        raw (None | bool | Auto): Whether to cache or delete the ``raw`` phase of the file, if its
+            format uses it. If ``bool``, it overrides what the intended use cases would do. If
+            ``None``, it takes its value from from the intended use cases would do. If ``Auto``,
+            falls back to ``None``.
+        can (None | bool | Auto): Whether to cache or delete the ``can`` phase of the file, if its
+            format uses it. If ``bool``, it overrides what the intended use cases would do. If
+            ``None``, it takes its value from from the intended use cases would do. If ``Auto``,
+            falls back to ``None``.
+    """
+
+    def __init__(
+            self,
+            storage: Union[bool, Auto] = Auto(),
+            intermediates: Union[bool, Auto] = Auto(),
+            checked_resumption: Union[bool, Auto] = Auto(),
+            access: Union[bool, Auto] = Auto(),
+            zip: Union[None, bool, Auto] = Auto(),
+            raw: Union[None, bool, Auto] = Auto(),
+            can: Union[None, bool, Auto] = Auto(),
+    ) -> None:
+        self.storage = storage
+        self.intermediates = intermediates
+        self.checked_resumption = checked_resumption
+        self.access = access
+        self.zip = zip
+        self.raw = raw
+        self.can = can
+
+    def to_safe(self) -> Self:
+        """Get the safe version of this Phaser, which never deletes the storage/download phase.
+
+        Returns:
+            Self: Safe version of this Phaser.
+        """
+        ret = deepcopy(self)
+        ret.storage = True
+        return ret
+
+    def get_phase_deletions(self, phase_locs: NDArray[np.int64]) -> NDArray[np.int64]:
+        """Get phase deletions.
+
+        Args:
+            phase_locs (NDArray[np.int64]): Phase localities.
+
+        Returns:
+            NDArray[np.int64]: Phase deletions.
+        """
+        return np.zeros(3, np.int64)  # TODO
```

Comment:
what's this?

Reply:
Whether to delete (0 = false) x 3 phases (Zip, Raw, Can)

“Out of time and risk, just hardcode it to cache everything for now and actually implement this later”

Reply:
Using ndarray here because I’m doing numpy operations on the returned triples out of convenience and clear code… although that code was probably mostly cut from this PR